Delete bindings when deleting a service instance

jeff-phillips-18 opened the pull request:
Fixes #2149

jeff-phillips-18 commented:
![image](https://user-images.githubusercontent.com/11633780/30766855-ec16dfba-9fc5-11e7-9193-91289e7a52d3.png)

![image](https://user-images.githubusercontent.com/11633780/30766867-065b78d6-9fc6-11e7-841c-1b3dda436e6c.png)

spadgett commented on app/scripts/services/serviceInstances.js:
```diff
@@ -3,64 +3,105 @@
+      };
+
+      if (CatalogService.SERVICE_CATALOG_ENABLED) {
+        DataService.list(resource, context, function (bindings) {
```
I believe everywhere we call this, we already have a list of the bindings. I'd rather just pass those in if we have them.

jeff-phillips-18 replied:
The service instance page does not have the bindings. I can take them optionally and only retrieve if not given.

spadgett replied:
> The service instance page does not have the bindings.

Hey @jeff-phillips-18 I'm OK with it in that case.

I want to look at the instance page anyway later. It looks like it's requesting service classes twice, and that's potentially a lot of data.

In general, I think it's better for the page controllers to get the data and pass them down to the components to avoid requesting things more than once.

jeff-phillips-18 replied:
I can take a look at that as part of this. Basically its the ResourceServiceBindings that gets the bindings for the service instance page (an all the 'application' pages. I can update that logic as well if you'd like.

spadgett replied:
It might be better to wait for #2155 since I have some changes to instance page and resource service bindings component in flight :/

jeff-phillips-18 replied:
Pushed changes to only get bindings when not passed. Holding off on updating all the controllers using the ResourceServiceBindings.

spadgett commented on app/scripts/services/serviceInstances.js:
```diff
@@ -3,64 +3,105 @@
+      });
+    };
+
+    var deprovision = function (apiObject) {
```
I realize this isn't new, but I'd rather call the parameter `serviceInstance` because the function is very specific to instances. `apiObject` sounds like any object.

spadgett commented on app/views/modals/warning.html:
```diff
@@ -0,0 +1,14 @@
+<div class="modal-resource-action">
+  <div class="modal-body">
+    <h1>
+      <span class="pficon pficon-warning-triangle-o"></span>
```
```html
<span class="pficon pficon-warning-triangle-o" aria-hidden="true"></span>
<span class="sr-only">warning</span>
```

spadgett commented on app/scripts/services/serviceInstances.js:
```diff
@@ -3,64 +3,105 @@
+        // TODO - remove once this is resolved https://github.com/kubernetes-incubator/service-catalog/issues/942
+        var opts = {
+          propagationPolicy: null
+        };
```
Not related to this PR, but we should double-check that this isn't needed for bindings as well. Right now we're using `delete-link` for bindings and in some cases and not passing `propogationPolicy: null`

spadgett replied:
Disregard, looks like we are handling it: https://github.com/openshift/origin-web-console/blob/master/app/scripts/directives/deleteLink.js#L153

spadgett commented on app/scripts/services/serviceInstances.js:
```diff
@@ -3,64 +3,105 @@
+        resolve: {
+          modalConfig: function() {
+            return {
+              title: "Unable to Deprovision",
```
We call the action "Delete" in the UI, so I'd be consistent in the messages we use (in the dialog and notifications).

ncameronbritt commented:
The original idea was to block the user from deleting until the bindings had been deleted, but I'm not sure we considered the possibility that deleting the bindings could take some time. 

I think it would be a better experience to allow users to delete the bindings from that dialog, rather than just telling them to go do it. The bindings would be then marked for deletion and the deprovision could go ahead as @spadgett suggested. Maybe something like this?
<img width="694" alt="screen shot 2017-09-23 at 10 10 15 am" src="https://user-images.githubusercontent.com/8974593/30773885-eaae6878-a047-11e7-9032-ffec12a740f3.png">

Related but probably outside of this scope, should we even be showing things that are marked for deletion in the UI? I know it doesn't work this way in practice, but deleting seems like an action that should be able to happen more or less instantaneously. Do users need to care about things they deleted but which are just hanging around waiting for the system to delete them?

ncameronbritt commented:
That "delete all bindings" checkbox should probably say "delete all
bindings for $service"


On Sat, Sep 23, 2017 at 10:20 AM Cameron Britt <notifications@github.com>
wrote:

> The original idea was to block the user from deleting until the bindings
> had been deleted, but I'm not sure we considered the possibility that
> deleting the bindings could take some time.
>
> I think it would be a better experience to allow users to delete the
> bindings from that dialog, rather than just telling them to go do it. The
> bindings would be then marked for deletion and the deprovision could go
> ahead as @spadgett <https://github.com/spadgett> suggested. Maybe
> something like this?
> [image: screen shot 2017-09-23 at 10 10 15 am]
> <https://user-images.githubusercontent.com/8974593/30773885-eaae6878-a047-11e7-9032-ffec12a740f3.png>
>
> Related but probably outside of this scope, should we even be showing
> things that are marked for deletion in the UI? I know it doesn't work this
> way in practice, but deleting seems like an action that should be able to
> happen more or less instantaneously. Do users need to care about things
> they deleted but which are just hanging around waiting for the system to
> delete them?
>
> —
> You are receiving this because you are subscribed to this thread.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/2150#issuecomment-331638828>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIjxASIzWiTUxFs8QzOLUKV3p1vhUjPZks5slRO4gaJpZM4PhUkh>
> .
>

jeff-phillips-18 commented:
@ncameronbritt How about the title?

jeff-phillips-18 commented:
@ncameronbritt @spadgett  How about something like:

![image](https://user-images.githubusercontent.com/11633780/30817601-45ce4eb2-a1e7-11e7-8e75-e763d2697a65.png)

ncameronbritt commented:
@spadgett and I were just talking about this. I think the best experience is to simply delete the bindings for the user. Deleting the service will also delete any deployments created by the service, and there's no reason for the user to think that bindings needed to be treated in a special way, and a binding doesn't have any value without the service.

Maybe it makes sense to add a sentence like we have [here](https://github.com/openshift/openshift-origin-design/blob/master/web-console/3-project-details/binding-in-project.md#delete-service): "$service and its data will no longer be available to your application."

jeff-phillips-18 commented:
Updated to automatically delete bindings when deleting a service instance (updated PR title as well).
Confirmation Dialog:
![image](https://user-images.githubusercontent.com/11633780/30821908-f575af10-a1f4-11e7-88b5-0cd7d1398672.png)

ncameronbritt commented:
Works for me. I would probably change "It cannot be undone" to "This cannot be undone." And in the sentence before it should be "*your* applications".

jeff-phillips-18 commented:
>  I would probably change "It cannot be undone" to "This cannot be undone." 

This is in the delete utility code (been there for a while). I can change it, but it would change for most deletion confirmation dialogs. @ncameronbritt Please advise.

jeff-phillips-18 commented:
@spadgett Also added checks to not show the "no bindable services" message when the object is either a service instance or it is marked for deletion. Removed the "Create Binding" link when the object is marked for deletion.

spadgett commented on app/scripts/services/serviceInstances.js:
```diff
@@ -2,65 +2,129 @@
+        });
+      }
+
+      return deferred.promise;
```
You can just return the promise from `DataService.list` instead of creating your own deferred.

```js
    return DataService.list(resource, context, function (serviceBindings) {
      bindings = serviceBindings.by('metadata.name');
      return BindingService.getBindingsForResource(bindings, apiObject);
    });
```

jeff-phillips-18 replied:
@spadgett What comes back from the promise is not the bindings manipulated by the callback function. Maybe I'm doing something else wrong?

spadgett replied:
Sorry my JS was wrong:

```js
   return DataService.list(resource, context).then(function(serviceBindings) {
      bindings = serviceBindings.by('metadata.name');
      return BindingService.getBindingsForResource(bindings, apiObject);
    });
```

jeff-phillips-18 replied:
Ah yes, that makes sense. Sorry, I should have seen that.

jeff-phillips-18 replied:
Fixed. Thanks @spadgett

spadgett commented on app/scripts/services/serviceInstances.js:
```diff
@@ -2,65 +2,129 @@
+
+      getBindingsIfNecessary(apiObject, bindings).then(function(serviceBindings) {
+        _.each(serviceBindings, function (binding) {
+          if (!binding.metadata.deletionTimestamp) {
```
This is a style nit, but I personally find early returns easier to reason about.

```js
if (binding.metadata.deletionTimestamp) {
  // The binding is marked for deletion.
  return;
}

// <code to delete the binding>
```

jeff-phillips-18 replied:
fixed

ncameronbritt commented:
@jeff-phillips-18 I don't feel that strongly about "this" vs. "it", so if it's going to change a bunch of other stuff, I'd say leave it as is for now, and we can address the language change taking into account all of the different contexts in which it occurs later, if we feel like we need to.

jeff-phillips-18 commented on app/scripts/services/serviceInstances.js:
```diff
@@ -2,65 +2,129 @@
+        var resource = {
+          group: 'servicecatalog.k8s.io',
+          resource: 'serviceinstancecredentials'
+        };
```
fixed

jeff-phillips-18 commented on app/scripts/services/serviceInstances.js:
```diff
@@ -2,65 +2,129 @@
+      var resource = {
+        group: 'servicecatalog.k8s.io',
+        resource: 'serviceinstances'
+      };
```
fixed

jeff-phillips-18 commented on app/scripts/services/serviceInstances.js:
```diff
@@ -2,65 +2,129 @@
+      var resource = {
+        group: 'servicecatalog.k8s.io',
+        resource: 'serviceinstancecredentials'
+      };
```
fixed

spadgett commented:
[merge]
